
  <h1><center>Analyse des horaires de travail des employés</center></h1>

  <p>Dans ce notebook, nous allons explorer et analyser les données d'horaires de travail collectées à partir du système de badge de l'entreprise. La société utilise des lecteurs de badge depuis plusieurs années pour suivre les heures d'entrée et de sortie des employés. La direction a décidé de partager les heures d'entrée et de sortie de tous les employés sur une période sélectionnée de l'année, représentant un niveau d'activité moyen à travers tous les départements.</p>

  <h2>Objectif</h2>

  <p>L'objectif principal de cette analyse est d'obtenir des informations sur les habitudes de travail des employés, d'identifier les tendances ou anomalies dans leurs horaires de travail, et d'extraire des informations significatives qui pourraient contribuer à une meilleure gestion de la main-d'œuvre. En examinant les données allant du 1er janvier au 31 décembre 2015, nous visons à comprendre la distribution des heures de travail, explorer les périodes d'activité maximale et éventuellement corréler les données avec d'autres facteurs pertinents.</p>

  <h2>Description des données</h2>

  <p>Le jeu de données se compose de deux fichiers, chacun contenant les enregistrements des heures d'arrivée et de départ des employés par date. Ces horodatages devraient fournir une vue complète des routines quotidiennes des employés tout au long de l'année spécifiée. L'analyse impliquera le traitement, le nettoyage et la visualisation des données pour tirer des enseignements exploitables.</p>

  <p>Commençons par charger les données et effectuer une exploration initiale pour comprendre leur structure et leurs caractéristiques.</p>

### **Import des bibliothèques**

In [45]:
import pandas as pd
from work_tools import * # Ici se trouvent nos propres fonctions

### **Import et Consult des Fichiers**

In [46]:
# Charger les données d'horaires d'arrivée depuis le fichier CSV 'in_time.csv'
in_time = pd.read_csv('./in_time.csv')

# Charger les données d'horaires de départ depuis le fichier CSV 'out_time.csv'
out_time = pd.read_csv('./out_time.csv')

In [47]:
# Afficher les dix premières lignes du DataFrame 'in_time'
in_time.head(10)

Unnamed: 0  2015-01-01           2015-01-02           2015-01-05  \
0           1         NaN  2015-01-02 09:43:45  2015-01-05 10:08:48   
1           2         NaN  2015-01-02 10:15:44  2015-01-05 10:21:05   
2           3         NaN  2015-01-02 10:17:41  2015-01-05 09:50:50   
3           4         NaN  2015-01-02 10:05:06  2015-01-05 09:56:32   
4           5         NaN  2015-01-02 10:28:17  2015-01-05 09:49:58   
5           6         NaN  2015-01-02 09:43:08  2015-01-05 10:14:00   
6           7         NaN  2015-01-02 10:20:13  2015-01-05 09:30:01   
7           8         NaN  2015-01-02 09:57:10  2015-01-05 09:48:56   
8           9         NaN                  NaN  2015-01-05 10:01:42   
9          10         NaN  2015-01-02 09:55:53  2015-01-05 10:21:06   

            2015-01-06           2015-01-07           2015-01-08  \
0  2015-01-06 09:54:26  2015-01-07 09:34:31  2015-01-08 09:51:09   
1                  NaN  2015-01-07 09:45:17  2015-01-08 10:09:04   
2  2015-01-06 10:14:13  2015-01-07 09:47:27  2015-01-08 10:03:40   
3  2015-01-06 10:11:07  2015-01-07 09:37:30  2015-01-08 10:02:08   
4  2015-01-06 09:45:28  2015-01-07 09:49:37  2015-01-08 10:19:44   
5  2015-01-06 10:08:42  2015-01-07 10:18:15  2015-01-08 10:33:09   
6  2015-01-06 09:48:47  2015-01-07 09:46:18  2015-01-08 09:59:29   
7  2015-01-06 09:54:04  2015-01-07 09:52:31  2015-01-08 09:58:31   
8  2015-01-06 09:50:56  2015-01-07 10:02:57  2015-01-08 10:07:22   
9  2015-01-06 10:03:01  2015-01-07 10:06:01  2015-01-08 09:52:25   

            2015-01-09           2015-01-12           2015-01-13  ...  \
0  2015-01-09 10:09:25  2015-01-12 09:42:53  2015-01-13 10:13:06  ...   
1  2015-01-09 09:43:26  2015-01-12 10:00:07  2015-01-13 10:43:29  ...   
2  2015-01-09 10:05:49  2015-01-12 10:03:47  2015-01-13 10:21:26  ...   
3  2015-01-09 10:08:12  2015-01-12 10:13:42  2015-01-13 09:53:22  ...   
4  2015-01-09 10:00:50  2015-01-12 10:29:27  2015-01-13 09:59:32  ...   
5  2015-01-09 10:19:13  2015-01-12 09:48:30  2015-01-13 09:54:26  ...   
6  2015-01-09 10:13:26  2015-01-12 09:23:42  2015-01-13 10:00:14  ...   
7  2015-01-09 09:53:12  2015-01-12 09:58:57  2015-01-13 09:42:05  ...   
8  2015-01-09 09:59:54  2015-01-12 10:14:55                  NaN  ...   
9  2015-01-09 09:36:12  2015-01-12 09:59:26  2015-01-13 10:08:33  ...   

            2015-12-18           2015-12-21           2015-12-22  \
0                  NaN  2015-12-21 09:55:29  2015-12-22 10:04:06   
1  2015-12-18 10:37:17  2015-12-21 09:49:02  2015-12-22 10:33:51   
2  2015-12-18 10:15:14  2015-12-21 10:10:28  2015-12-22 09:44:44   
3  2015-12-18 10:17:38  2015-12-21 09:58:21  2015-12-22 10:04:25   
4  2015-12-18 09:58:35  2015-12-21 10:03:41  2015-12-22 10:10:30   
5  2015-12-18 10:24:55  2015-12-21 09:44:43  2015-12-22 09:38:00   
6  2015-12-18 10:12:15  2015-12-21 09:52:10                  NaN   
7  2015-12-18 10:16:27  2015-12-21 10:12:52  2015-12-22 10:58:57   
8  2015-12-18 10:19:42  2015-12-21 10:18:09  2015-12-22 09:39:54   
9  2015-12-18 09:57:20  2015-12-21 09:39:07  2015-12-22 10:17:05   

            2015-12-23           2015-12-24 2015-12-25           2015-12-28  \
0  2015-12-23 10:14:27  2015-12-24 10:11:35        NaN  2015-12-28 10:13:41   
1  2015-12-23 10:12:10                  NaN        NaN  2015-12-28 09:31:45   
2  2015-12-23 10:15:54  2015-12-24 10:07:26        NaN  2015-12-28 09:42:05   
3  2015-12-23 10:11:46  2015-12-24 09:43:15        NaN  2015-12-28 09:52:44   
4  2015-12-23 10:13:36  2015-12-24 09:44:24        NaN  2015-12-28 10:05:15   
5  2015-12-23 09:53:27  2015-12-24 09:38:46        NaN                  NaN   
6  2015-12-23 10:22:03  2015-12-24 10:24:29        NaN  2015-12-28 10:19:47   
7  2015-12-23 09:45:12  2015-12-24 09:46:29        NaN  2015-12-28 09:59:24   
8  2015-12-23 09:48:05  2015-12-24 10:00:32        NaN                  NaN   
9  2015-12-23 10:25:33  2015-12-24 10:21:01        NaN  2015-12-28 10:20:31   

            2015-12-29           2015-12-30      

In [48]:
# Afficher les dix premières lignes du DataFrame 'out_time'
out_time.head(10)

Unnamed: 0  2015-01-01           2015-01-02           2015-01-05  \
0           1         NaN  2015-01-02 16:56:15  2015-01-05 17:20:11   
1           2         NaN  2015-01-02 18:22:17  2015-01-05 17:48:22   
2           3         NaN  2015-01-02 16:59:14  2015-01-05 17:06:46   
3           4         NaN  2015-01-02 17:25:24  2015-01-05 17:14:03   
4           5         NaN  2015-01-02 18:31:37  2015-01-05 17:49:15   
5           6         NaN  2015-01-02 20:29:54  2015-01-05 20:57:19   
6           7         NaN  2015-01-02 17:10:31  2015-01-05 17:02:19   
7           8         NaN  2015-01-02 17:02:35  2015-01-05 16:52:09   
8           9         NaN                  NaN  2015-01-05 17:00:43   
9          10         NaN  2015-01-02 17:17:31  2015-01-05 17:27:11   

            2015-01-06           2015-01-07           2015-01-08  \
0  2015-01-06 17:19:05  2015-01-07 16:34:55  2015-01-08 17:08:32   
1                  NaN  2015-01-07 17:09:06  2015-01-08 17:34:04   
2  2015-01-06 16:38:32  2015-01-07 16:33:21  2015-01-08 17:24:22   
3  2015-01-06 17:07:42  2015-01-07 16:32:40  2015-01-08 16:53:11   
4  2015-01-06 17:26:25  2015-01-07 17:37:59  2015-01-08 17:59:28   
5  2015-01-06 21:06:31  2015-01-07 20:36:10  2015-01-08 21:33:43   
6  2015-01-06 17:04:47  2015-01-07 16:11:37  2015-01-08 17:01:52   
7  2015-01-06 16:33:13  2015-01-07 16:42:05  2015-01-08 16:18:14   
8  2015-01-06 17:10:01  2015-01-07 17:36:23  2015-01-08 17:30:35   
9  2015-01-06 17:33:55  2015-01-07 17:15:12  2015-01-08 16:42:03   

            2015-01-09           2015-01-12           2015-01-13  ...  \
0  2015-01-09 17:38:29  2015-01-12 16:58:39  2015-01-13 18:02:58  ...   
1  2015-01-09 16:52:29  2015-01-12 17:36:48  2015-01-13 18:00:13  ...   
2  2015-01-09 16:57:30  2015-01-12 17:28:54  2015-01-13 17:21:25  ...   
3  2015-01-09 17:19:47  2015-01-12 17:13:37  2015-01-13 17:11:45  ...   
4  2015-01-09 17:44:08  2015-01-12 18:51:21  2015-01-13 18:14:58  ...   
5  2015-01-09 21:25:12  2015-01-12 20:38:47  2015-01-13 20:10:38  ...   
6  2015-01-09 17:23:17  2015-01-12 16:04:41  2015-01-13 16:55:05  ...   
7  2015-01-09 16:49:56  2015-01-12 16:41:14  2015-01-13 16:26:12  ...   
8  2015-01-09 17:19:58  2015-01-12 17:14:54                  NaN  ...   
9  2015-01-09 16:37:48  2015-01-12 17:54:23  2015-01-13 17:21:32  ...   

            2015-12-18           2015-12-21           2015-12-22  \
0                  NaN  2015-12-21 17:15:50  2015-12-22 17:27:51   
1  2015-12-18 18:31:28  2015-12-21 17:34:16  2015-12-22 18:16:35   
2  2015-12-18 17:02:23  2015-12-21 17:20:17  2015-12-22 16:32:50   
3  2015-12-18 17:55:23  2015-12-21 16:49:09  2015-12-22 17:24:00   
4  2015-12-18 17:52:48  2015-12-21 17:43:35  2015-12-22 18:07:57   
5  2015-12-18 20:58:47  2015-12-21 20:48:45  2015-12-22 20:46:45   
6  2015-12-18 17:39:13  2015-12-21 16:29:19                  NaN   
7  2015-12-18 17:20:14  2015-12-21 17:16:25  2015-12-22 17:37:19   
8  2015-12-18 16:54:59  2015-12-21 17:29:59  2015-12-22 16:58:51   
9  2015-12-18 17:20:04  2015-12-21 16:46:26  2015-12-22 17:38:33   

            2015-12-23           2015-12-24 2015-12-25           2015-12-28  \
0  2015-12-23 16:44:44  2015-12-24 17:47:22        NaN  2015-12-28 18:00:07   
1  2015-12-23 17:38:18                  NaN        NaN  2015-12-28 17:08:38   
2  2015-12-23 16:59:43  2015-12-24 16:58:25        NaN  2015-12-28 16:43:31   
3  2015-12-23 17:36:35  2015-12-24 16:48:21        NaN  2015-12-28 17:19:34   
4  2015-12-23 18:00:49  2015-12-24 17:59:22        NaN  2015-12-28 17:44:59   
5  2015-12-23 20:51:06  2015-12-24 20:06:50        NaN                  NaN   
6  2015-12-23 17:19:40  2015-12-24 17:19:19        NaN  2015-12-28 16:52:09   
7  2015-12-23 16:33:44  2015-12-24 16:23:30        NaN  2015-12-28 16:16:24   
8  2015-12-23 17:10:38  2015-12-24 16:47:09        NaN                  NaN   
9  2015-12-23 17:14:07  2015-12-24 17:40:42        NaN  2015-12-28 17:18:11   

            2015-12-29           2015-12-30      

In [49]:
# Afficher les dimensions du DataFrame 'in_time' (nombre de lignes et de colonnes)
print(in_time.shape)

# Afficher les dimensions du DataFrame 'out_time' (nombre de lignes et de colonnes)
out_time.shape

(4410, 262)


(4410, 262)

  <p>Les dimensions (shape) (4410, 251), représentent le nombre de lignes et de colonnes du DataFrame. En particulier :</p>

  <ul>
    <li>Pour le DataFrame 'in_time': Il comporte 4410 lignes et 251 colonnes.</li>
    <li>Pour le DataFrame 'out_time': Il comporte également 4410 lignes et 251 colonnes.</li>
  </ul>

  <p>Ces nombres indiquent que chaque DataFrame contient 4410 enregistrements (entrées ou observations) et 251 variables (attributs ou colonnes).</p>

In [50]:
# Afficher le nombre de valeurs nulles par colonne dans le DataFrame 'in_time'
in_time.isna().sum()

Unnamed: 0       0
2015-01-01    4410
2015-01-02     209
2015-01-05     206
2015-01-06     228
              ... 
2015-12-25    4410
2015-12-28     234
2015-12-29     230
2015-12-30     265
2015-12-31     213
Length: 262, dtype: int64

In [51]:
# Afficher le nombre de valeurs nulles par colonne dans le DataFrame 'out_time'
out_time.isna().sum()

Unnamed: 0       0
2015-01-01    4410
2015-01-02     209
2015-01-05     206
2015-01-06     228
              ... 
2015-12-25    4410
2015-12-28     234
2015-12-29     230
2015-12-30     265
2015-12-31     213
Length: 262, dtype: int64

<pre>
    in_time.isna().sum()
    out_time.isna().sum()
</pre>

  <p>
    Ces sorties mettent en évidence que plusieurs attributs du DataFrame 'in_time' et 'out_time' contiennent des valeurs nulles.
    La présence de valeurs nulles dans un ensemble de données peut nécessiter une attention particulière lors du traitement et de l'analyse ultérieure.
  </p>

In [52]:
# Vérifier les doublons dans le DataFrame 'in_time'
check_for_duplicates(in_time)

print('-------------')

# Vérifier les doublons dans le DataFrame 'out_time'
check_for_duplicates(out_time)

Empty DataFrame
Columns: [Unnamed: 0, 2015-01-01, 2015-01-02, 2015-01-05, 2015-01-06, 2015-01-07, 2015-01-08, 2015-01-09, 2015-01-12, 2015-01-13, 2015-01-14, 2015-01-15, 2015-01-16, 2015-01-19, 2015-01-20, 2015-01-21, 2015-01-22, 2015-01-23, 2015-01-26, 2015-01-27, 2015-01-28, 2015-01-29, 2015-01-30, 2015-02-02, 2015-02-03, 2015-02-04, 2015-02-05, 2015-02-06, 2015-02-09, 2015-02-10, 2015-02-11, 2015-02-12, 2015-02-13, 2015-02-16, 2015-02-17, 2015-02-18, 2015-02-19, 2015-02-20, 2015-02-23, 2015-02-24, 2015-02-25, 2015-02-26, 2015-02-27, 2015-03-02, 2015-03-03, 2015-03-04, 2015-03-05, 2015-03-06, 2015-03-09, 2015-03-10, 2015-03-11, 2015-03-12, 2015-03-13, 2015-03-16, 2015-03-17, 2015-03-18, 2015-03-19, 2015-03-20, 2015-03-23, 2015-03-24, 2015-03-25, 2015-03-26, 2015-03-27, 2015-03-30, 2015-03-31, 2015-04-01, 2015-04-02, 2015-04-03, 2015-04-06, 2015-04-07, 2015-04-08, 2015-04-09, 2015-04-10, 2015-04-13, 2015-04-14, 2015-04-15, 2015-04-16, 2015-04-17, 2015-04-20, 2015-04-21, 2015-04-22, 20

<p>
    On a bien indiqué qu'aucun doublon n'a été trouvé dans le DataFrame 'out_time'. 
    Comme pour 'in_time', l'absence de doublons est une caractéristique positive de la qualité des données.
  </p>

### **Prétraitement des Données**

Sélection des Colonnes à Supprimer et à Manipuler avec des Valeurs Nulles

In [53]:
# Récupérer les colonnes à supprimer et les colonnes nécessitant une manipulation des valeurs nulles dans 'in_time'
cols_to_drop, cols_to_manipulate_nan = fetch_date_time_nan_columns(in_time)

Vérification des Valeurs Nulles dans les Colonnes à Supprimer

In [54]:
# Afficher le nombre de valeurs nulles pour les colonnes à supprimer dans 'in_time'
in_time[cols_to_drop].isna().sum()

2015-01-01    4410
2015-01-14    4410
2015-01-26    4410
2015-03-05    4410
2015-05-01    4410
2015-07-17    4410
2015-09-17    4410
2015-10-02    4410
2015-11-09    4410
2015-11-10    4410
2015-11-11    4410
2015-12-25    4410
dtype: int64

Vérification des Valeurs Nulles dans les Colonnes à Manipuler

In [55]:
# Afficher le nombre de valeurs nulles pour les colonnes nécessitant une manipulation des valeurs nulles dans 'in_time'
in_time[cols_to_manipulate_nan].isna().sum()

2015-01-02    209
2015-01-05    206
2015-01-06    228
2015-01-07    209
2015-01-08    243
             ... 
2015-12-24    250
2015-12-28    234
2015-12-29    230
2015-12-30    265
2015-12-31    213
Length: 249, dtype: int64

Suppression des Colonnes Indésirables

In [56]:
# Supprimer les colonnes indésirables dans 'in_time'
in_time = drop_columns(in_time, cols_to_drop)

# Afficher le DataFrame 'in_time' après la suppression des colonnes
in_time

Unnamed: 0           2015-01-02           2015-01-05  \
0              1  2015-01-02 09:43:45  2015-01-05 10:08:48   
1              2  2015-01-02 10:15:44  2015-01-05 10:21:05   
2              3  2015-01-02 10:17:41  2015-01-05 09:50:50   
3              4  2015-01-02 10:05:06  2015-01-05 09:56:32   
4              5  2015-01-02 10:28:17  2015-01-05 09:49:58   
...          ...                  ...                  ...   
4405        4406  2015-01-02 09:20:32  2015-01-05 10:17:53   
4406        4407  2015-01-02 10:03:41                  NaN   
4407        4408  2015-01-02 10:01:01  2015-01-05 09:33:00   
4408        4409  2015-01-02 10:17:05  2015-01-05 10:02:27   
4409        4410  2015-01-02 09:59:09  2015-01-05 10:16:14   

               2015-01-06           2015-01-07           2015-01-08  \
0     2015-01-06 09:54:26  2015-01-07 09:34:31  2015-01-08 09:51:09   
1                     NaN  2015-01-07 09:45:17  2015-01-08 10:09:04   
2     2015-01-06 10:14:13  2015-01-07 09:47:27  2015-01-08 10:03:40   
3     2015-01-06 10:11:07  2015-01-07 09:37:30  2015-01-08 10:02:08   
4     2015-01-06 09:45:28  2015-01-07 09:49:37  2015-01-08 10:19:44   
...                   ...                  ...                  ...   
4405  2015-01-06 10:26:51  2015-01-07 10:06:58  2015-01-08 09:45:06   
4406  2015-01-06 09:44:00  2015-01-07 09:42:10  2015-01-08 10:00:57   
4407  2015-01-06 09:49:17  2015-01-07 10:28:12  2015-01-08 09:47:38   
4408  2015-01-06 10:12:50  2015-01-07 10:12:31  2015-01-08 09:42:57   
4409  2015-01-06 09:52:30  2015-01-07 09:43:15  2015-01-08 10:06:55   

               2015-01-09           2015-01-12           2015-01-13  \
0     2015-01-09 10:09:25  2015-01-12 09:42:53  2015-01-13 10:13:06   
1     2015-01-09 09:43:26  2015-01-12 10:00:07  2015-01-13 10:43:29   
2     2015-01-09 10:05:49  2015-01-12 10:03:47  2015-01-13 10:21:26   
3     2015-01-09 10:08:12  2015-01-12 10:13:42  2015-01-13 09:53:22   
4     2015-01-09 10:00:50  2015-01-12 10:29:27  2015-01-13 09:59:32   
...                   ...                  ...                  ...   
4405  2015-01-09 09:49:24  2015-01-12 09:37:10  2015-01-13 09:25:02   
4406  2015-01-09 09:44:04  2015-01-12 10:07:32  2015-01-13 10:05:11   
4407  2015-01-09 10:01:03  2015-01-12 09:49:12  2015-01-13 09:47:10   
4408                  NaN  2015-01-12 10:00:38  2015-01-13 09:48:03   
4409  2015-01-09 10:27:39  2015-01-12 09:47:35  2015-01-13 09:30:00   

               2015-01-15  ...           2015-12-17           2015-12-18  \
0     2015-01-15 10:01:24  ...                  NaN                  NaN   
1     2015-01-15 09:37:57  ...  2015-12-17 09:15:08  2015-12-18 10:37:17   
2     2015-01-15 09:55:11  ...  2015-12-17 09:53:17  2015-12-18 10:15:14   
3     2015-01-15 10:00:50  ...  2015-12-17 09:54:36  2015-12-18 10:17:38   
4     2015-01-15 10:06:12  ...  2015-12-17 09:46:35  2015-12-18 09:58:35   
...                   ...  ...                  ...                  ...   
4405  2015-01-15 09:29:17  ...  2015-12-17 10:05:22  2015-12-18 10:01:06   
4406  2015-01-15 10:18:11  ...  2015-12-17 09:45:49  2015-12-18 09:27:32   
4407  2015-01-15 10:08:31  ...  2015-12-17 10:01:36  2015-12-18 10:00:57   
4408  2015-01-15 09:04:17  ...  2015-12-17 09:51:40  2015-12-18 09:54:33   
4409  2015-01-15 10:08:19  ...  2015-12-17 09:58:17  2015-12-18 10:20:40   

               2015-12-21           2015-12-22           2015-12-23  \
0     2015-12-21 09:55:29  2015-12-22 10:04:06  2015-12-23 10:14:27   
1     2015-12-21 09:49:02  2015-12-22 10:33:51  2015-12-23 10:12:10   
2     2015-12-21 10:10:28  2015-12-22 09:44:44  2015-12-23 10:15:54   
3     2015-12-21 09:58:21  2015-12-22 10:04:25  2015-12-23 10:11:46   
4     2015-12-21 10:03:41  2015-12-22 10:10:30  2015-12-23 10:13:36   
...                   ...                  ...                  ...   
4405  2015-12-21 10:25:25  2015-12-22 10:16:11  2015-12-23 10:04:40   
4406  2015-12-21 09:41:24  2015-12-22 09:50:30  2015-12-23 10:32:21   
440

<p>
    Les colonnes indésirables, principalement remplies de valeurs nulles, ont été avec succès supprimées du DataFrame 'in_time'. 
    À présent, nous allons aborder les colonnes restantes qui contiennent encore certaines valeurs nulles.
  </p>

In [57]:
# Afficher le nombre de valeurs nulles par colonne dans le DataFrame 'in_time'
in_time.isna().sum()

Unnamed: 0      0
2015-01-02    209
2015-01-05    206
2015-01-06    228
2015-01-07    209
             ... 
2015-12-24    250
2015-12-28    234
2015-12-29    230
2015-12-30    265
2015-12-31    213
Length: 250, dtype: int64

Interpolation des Valeurs Nulles dans les Colonnes à Manipuler

In [58]:
# Interpoler les valeurs nulles dans les colonnes nécessitant une manipulation des valeurs nulles dans 'in_time'
in_time = interpolate_datetime_nulls(in_time, cols_to_manipulate_nan)

e:\Sami\Sami\Sami\ISIMa\2 TSD\Data_Analysis_Python\HumanForYou\python_project\python_project\work_tools.py:296: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  dataframe[column_name] = dataframe[column_name].fillna(method='bfill')
e:\Sami\Sami\Sami\ISIMa\2 TSD\Data_Analysis_Python\HumanForYou\python_project\python_project\work_tools.py:299: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  dataframe[column_name] = dataframe[column_name].fillna(method='ffill')


In [59]:
in_time.isna().sum()

Unnamed: 0    0
2015-01-02    0
2015-01-05    0
2015-01-06    0
2015-01-07    0
             ..
2015-12-24    0
2015-12-28    0
2015-12-29    0
2015-12-30    0
2015-12-31    0
Length: 250, dtype: int64

Super! Aucune valeur nulle existe dans la dataset

In [60]:
in_time.sample(10)

Unnamed: 0          2015-01-02          2015-01-05          2015-01-06  \
2149        2150 2015-01-02 09:55:14 2015-01-05 09:40:30 2015-01-06 09:42:28   
3972        3973 2015-01-02 10:04:21 2015-01-05 09:38:11 2015-01-06 10:15:25   
3952        3953 2015-01-02 09:58:50 2015-01-05 09:47:54 2015-01-06 10:07:35   
1024        1025 2015-01-02 10:19:04 2015-01-05 10:16:31 2015-01-06 10:20:05   
3940        3941 2015-01-02 09:49:54 2015-01-05 10:16:14 2015-01-06 10:18:46   
2123        2124 2015-01-02 09:41:36 2015-01-05 10:22:58 2015-01-06 10:40:59   
3214        3215 2015-01-02 10:24:39 2015-01-05 10:14:20 2015-01-06 09:46:55   
2415        2416 2015-01-02 10:21:57 2015-01-05 09:48:17 2015-01-06 09:58:13   
3507        3508 2015-01-02 09:56:25 2015-01-05 10:13:02 2015-01-06 10:02:11   
1510        1511 2015-01-02 10:00:34 2015-01-05 10:12:04 2015-01-06 10:39:22   

              2015-01-07          2015-01-08          2015-01-09  \
2149 2015-01-07 10:05:10 2015-01-08 09:50:46 2015-01-09 10:08:02   
3972 2015-01-07 09:53:01 2015-01-08 09:47:11 2015-01-09 09:43:56   
3952 2015-01-07 10:26:01 2015-01-08 10:07:47 2015-01-09 10:35:16   
1024 2015-01-07 09:44:59 2015-01-08 09:49:33 2015-01-09 10:01:46   
3940 2015-01-07 10:03:12 2015-01-08 09:45:16 2015-01-09 09:38:22   
2123 2015-01-07 10:16:43 2015-01-08 10:09:15 2015-01-09 10:10:23   
3214 2015-01-07 09:53:40 2015-01-08 09:50:00 2015-01-09 10:17:30   
2415 2015-01-07 09:21:15 2015-01-08 10:04:44 2015-01-09 10:15:29   
3507 2015-01-07 09:56:18 2015-01-08 10:11:18 2015-01-09 10:05:45   
1510 2015-01-07 10:04:40 2015-01-08 10:01:14 2015-01-09 09:59:55   

              2015-01-12          2015-01-13          2015-01-15  ...  \
2149 2015-01-12 09:43:17 2015-01-13 10:02:53 2015-01-15 10:14:22  ...   
3972 2015-01-12 10:47:55 2015-01-13 10:10:48 2015-01-15 09:13:10  ...   
3952 2015-01-12 10:01:43 2015-01-13 10:12:49 2015-01-15 10:23:03  ...   
1024 2015-01-12 09:56:34 2015-01-13 09:59:40 2015-01-15 09:54:44  ...   
3940 2015-01-12 10:20:35 2015-01-13 09:45:42 2015-01-15 10:08:11  ...   
2123 2015-01-12 09:52:10 2015-01-13 10:15:35 2015-01-15 09:54:37  ...   
3214 2015-01-12 10:10:06 2015-01-13 10:19:15 2015-01-15 10:18:32  ...   
2415 2015-01-12 10:00:56 2015-01-13 09:55:15 2015-01-15 09:48:15  ...   
3507 2015-01-12 10:04:24 2015-01-13 09:53:01 2015-01-15 10:02:14  ...   
1510 2015-01-12 10:12:26 2015-01-13 09:48:08 2015-01-15 09:38:04  ...   

              2015-12-17          2015-12-18          2015-12-21  \
2149 2015-12-17 09:56:50 2015-12-18 10:12:39 2015-12-21 09:37:41   
3972 2015-12-17 09:48:09 2015-12-18 10:16:16 2015-12-21 10:58:30   
3952 2015-12-17 09:49:22 2015-12-18 09:54:39 2015-12-21 09:43:55   
1024 2015-12-17 10:02:19 2015-12-18 09:22:16 2015-12-21 09:40:11   
3940 2015-12-17 09:47:36 2015-12-18 09:51:40 2015-12-21 10:09:52   
2123 2015-12-17 10:11:48 2015-12-18 09:29:27 2015-12-21 10:12:38   
3214 2015-12-17 09:46:33 2015-12-18 09:44:31 2015-12-21 09:51:10   
2415 2015-12-17 09:21:29 2015-12-18 10:05:19 2015-12-21 09:10:32   
3507 2015-12-17 10:35:29 2015-12-18 10:29:37 2015-12-21 09:30:59   
1510 2015-12-17 10:03:58 2015-12-18 09:37:43 2015-12-21 09:40:38   

              2015-12-22          2015-12-23          2015-12-24  \
2149 2015-12-22 10:19:12 2015-12-23 10:01:31 2015-12-24 10:27:13   
3972 2015-12-22 09:54:06 2015-12-23 10:11:08 2015-12-24 09:55:54   
3952 2015-12-22 10:17:07 2015-12-23 09:55:28 2015-12-24 09:52:13   
1024 2015-12-22 10:15:21 2015-12-23 10:17:52 2015-12-24 09:58:13   
3940 2015-12-22 10:07:45 2015-12-23 10:21:00 2015-12-24 09:34:41   
2123 2015-12-22 09:24:12 2015-12-23 09:30:38 2015-12-24 09:45:08   
3214 2015-12-22 09:52:46 2015-12-23 10:25:51 2015-12-24 10:26:17   
2415 2015-12-22 09:27:02 2015-12-23 10:15:55 2015-12-24 09:55:35   
3507 2015-12-22 10:05:51 2015-12-23 10:25:52 2015-12-24 09:31:16   
1510 2015-12-22 10:30:38 2015-12-23 09:51:55 2015-12-24 09:57:57   

              2015-12-28          2015-12-29          2015-12-30  \
2149 2

<p>Les étapes précédentes seront répétées pour le DataFrame 'out_time'.</p>

In [61]:
# Récupérer les colonnes à supprimer et les colonnes nécessitant une manipulation des valeurs nulles dans 'out_time'
cols_to_drop_out, cols_to_manipulate_nan_out = fetch_date_time_nan_columns(out_time)

In [62]:
# Afficher le nombre de valeurs nulles pour les colonnes à supprimer dans 'out_time'
out_time[cols_to_drop_out].isna().sum()

2015-01-01    4410
2015-01-14    4410
2015-01-26    4410
2015-03-05    4410
2015-05-01    4410
2015-07-17    4410
2015-09-17    4410
2015-10-02    4410
2015-11-09    4410
2015-11-10    4410
2015-11-11    4410
2015-12-25    4410
dtype: int64

In [63]:
# Afficher le nombre de valeurs nulles pour les colonnes nécessitant une manipulation des valeurs nulles dans 'out_time'
out_time[cols_to_manipulate_nan_out].isna().sum()

2015-01-02    209
2015-01-05    206
2015-01-06    228
2015-01-07    209
2015-01-08    243
             ... 
2015-12-24    250
2015-12-28    234
2015-12-29    230
2015-12-30    265
2015-12-31    213
Length: 249, dtype: int64

In [64]:
# Supprimer les colonnes indésirables dans 'out_time'
out_time = drop_columns(out_time, cols_to_drop_out)

# Afficher le DataFrame 'out_time' après la suppression des colonnes
out_time

Unnamed: 0           2015-01-02           2015-01-05  \
0              1  2015-01-02 16:56:15  2015-01-05 17:20:11   
1              2  2015-01-02 18:22:17  2015-01-05 17:48:22   
2              3  2015-01-02 16:59:14  2015-01-05 17:06:46   
3              4  2015-01-02 17:25:24  2015-01-05 17:14:03   
4              5  2015-01-02 18:31:37  2015-01-05 17:49:15   
...          ...                  ...                  ...   
4405        4406  2015-01-02 17:27:37  2015-01-05 19:08:20   
4406        4407  2015-01-02 16:19:01                  NaN   
4407        4408  2015-01-02 17:17:35  2015-01-05 17:08:07   
4408        4409  2015-01-02 19:48:37  2015-01-05 19:37:40   
4409        4410  2015-01-02 16:49:19  2015-01-05 17:33:02   

               2015-01-06           2015-01-07           2015-01-08  \
0     2015-01-06 17:19:05  2015-01-07 16:34:55  2015-01-08 17:08:32   
1                     NaN  2015-01-07 17:09:06  2015-01-08 17:34:04   
2     2015-01-06 16:38:32  2015-01-07 16:33:21  2015-01-08 17:24:22   
3     2015-01-06 17:07:42  2015-01-07 16:32:40  2015-01-08 16:53:11   
4     2015-01-06 17:26:25  2015-01-07 17:37:59  2015-01-08 17:59:28   
...                   ...                  ...                  ...   
4405  2015-01-06 18:50:49  2015-01-07 18:57:40  2015-01-08 17:58:31   
4406  2015-01-06 15:07:37  2015-01-07 15:25:50  2015-01-08 16:12:33   
4407  2015-01-06 17:27:46  2015-01-07 18:27:22  2015-01-08 17:05:25   
4408  2015-01-06 20:00:08  2015-01-07 19:35:59  2015-01-08 18:55:13   
4409  2015-01-06 16:36:10  2015-01-07 16:33:47  2015-01-08 17:32:31   

               2015-01-09           2015-01-12           2015-01-13  \
0     2015-01-09 17:38:29  2015-01-12 16:58:39  2015-01-13 18:02:58   
1     2015-01-09 16:52:29  2015-01-12 17:36:48  2015-01-13 18:00:13   
2     2015-01-09 16:57:30  2015-01-12 17:28:54  2015-01-13 17:21:25   
3     2015-01-09 17:19:47  2015-01-12 17:13:37  2015-01-13 17:11:45   
4     2015-01-09 17:44:08  2015-01-12 18:51:21  2015-01-13 18:14:58   
...                   ...                  ...                  ...   
4405  2015-01-09 18:06:15  2015-01-12 17:58:48  2015-01-13 18:10:35   
4406  2015-01-09 15:26:56  2015-01-12 16:10:42  2015-01-13 16:22:43   
4407  2015-01-09 17:02:57  2015-01-12 17:35:45  2015-01-13 17:15:52   
4408                  NaN  2015-01-12 19:18:17  2015-01-13 19:24:02   
4409  2015-01-09 17:25:58  2015-01-12 16:39:21  2015-01-13 16:59:28   

               2015-01-15  ...           2015-12-17           2015-12-18  \
0     2015-01-15 17:22:13  ...                  NaN                  NaN   
1     2015-01-15 17:14:44  ...  2015-12-17 17:10:50  2015-12-18 18:31:28   
2     2015-01-15 17:21:29  ...  2015-12-17 17:06:23  2015-12-18 17:02:23   
3     2015-01-15 16:53:26  ...  2015-12-17 17:21:57  2015-12-18 17:55:23   
4     2015-01-15 18:21:48  ...  2015-12-17 18:05:47  2015-12-18 17:52:48   
...                   ...  ...                  ...                  ...   
4405  2015-01-15 17:50:37  ...  2015-12-17 18:35:39  2015-12-18 18:06:05   
4406  2015-01-15 16:19:00  ...  2015-12-17 15:44:50  2015-12-18 15:23:02   
4407  2015-01-15 18:15:53  ...  2015-12-17 17:42:43  2015-12-18 17:48:05   
4408  2015-01-15 18:33:21  ...  2015-12-17 19:06:58  2015-12-18 19:52:44   
4409  2015-01-15 17:13:51  ...  2015-12-17 17:05:40  2015-12-18 17:23:49   

               2015-12-21           2015-12-22           2015-12-23  \
0     2015-12-21 17:15:50  2015-12-22 17:27:51  2015-12-23 16:44:44   
1     2015-12-21 17:34:16  2015-12-22 18:16:35  2015-12-23 17:38:18   
2     2015-12-21 17:20:17  2015-12-22 16:32:50  2015-12-23 16:59:43   
3     2015-12-21 16:49:09  2015-12-22 17:24:00  2015-12-23 17:36:35   
4     2015-12-21 17:43:35  2015-12-22 18:07:57  2015-12-23 18:00:49   
...                   ...                  ...                  ...   
4405  2015-12-21 18:35:06  2015-12-22 18:33:44  2015-12-23 18:40:56   
4406  2015-12-21 15:31:14  2015-12-22 15:45:59  2015-12-23 16:38:59   
440

In [65]:
out_time.isna().sum()

Unnamed: 0      0
2015-01-02    209
2015-01-05    206
2015-01-06    228
2015-01-07    209
             ... 
2015-12-24    250
2015-12-28    234
2015-12-29    230
2015-12-30    265
2015-12-31    213
Length: 250, dtype: int64

In [66]:
# Interpoler les valeurs nulles dans les colonnes nécessitant une manipulation des valeurs nulles dans 'out_time'
out_time = interpolate_datetime_nulls(out_time, cols_to_manipulate_nan_out)

e:\Sami\Sami\Sami\ISIMa\2 TSD\Data_Analysis_Python\HumanForYou\python_project\python_project\work_tools.py:296: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  dataframe[column_name] = dataframe[column_name].fillna(method='bfill')
e:\Sami\Sami\Sami\ISIMa\2 TSD\Data_Analysis_Python\HumanForYou\python_project\python_project\work_tools.py:299: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  dataframe[column_name] = dataframe[column_name].fillna(method='ffill')


In [67]:
out_time.isna().sum()

Unnamed: 0    0
2015-01-02    0
2015-01-05    0
2015-01-06    0
2015-01-07    0
             ..
2015-12-24    0
2015-12-28    0
2015-12-29    0
2015-12-30    0
2015-12-31    0
Length: 250, dtype: int64

In [68]:
out_time.sample(10)

Unnamed: 0          2015-01-02          2015-01-05          2015-01-06  \
1353        1354 2015-01-02 17:59:48 2015-01-05 16:46:37 2015-01-06 20:48:55   
994          995 2015-01-02 19:26:25 2015-01-05 19:27:36 2015-01-06 19:53:18   
1476        1477 2015-01-02 16:28:33 2015-01-05 16:41:37 2015-01-06 16:40:41   
3437        3438 2015-01-02 17:03:14 2015-01-05 17:34:28 2015-01-06 17:57:58   
4397        4398 2015-01-02 16:49:13 2015-01-05 17:48:35 2015-01-06 17:20:39   
813          814 2015-01-02 16:32:13 2015-01-05 16:33:28 2015-01-06 16:33:48   
4204        4205 2015-01-02 19:34:10 2015-01-05 19:34:43 2015-01-06 19:15:01   
2114        2115 2015-01-02 17:51:15 2015-01-05 18:32:39 2015-01-06 17:45:20   
3843        3844 2015-01-02 19:40:11 2015-01-05 20:01:18 2015-01-06 19:25:45   
4277        4278 2015-01-02 17:49:55 2015-01-05 16:46:01 2015-01-06 17:35:00   

              2015-01-07          2015-01-08          2015-01-09  \
1353 2015-01-07 19:49:33 2015-01-08 16:16:27 2015-01-09 18:00:26   
994  2015-01-07 19:57:06 2015-01-08 19:35:33 2015-01-09 19:40:25   
1476 2015-01-07 17:09:20 2015-01-08 17:17:52 2015-01-09 17:14:28   
3437 2015-01-07 17:13:02 2015-01-08 17:39:30 2015-01-09 16:53:23   
4397 2015-01-07 16:58:43 2015-01-08 17:46:57 2015-01-09 17:11:03   
813  2015-01-07 16:20:14 2015-01-08 16:09:57 2015-01-09 16:14:25   
4204 2015-01-07 19:02:28 2015-01-08 19:22:57 2015-01-09 18:46:13   
2114 2015-01-07 18:12:01 2015-01-08 17:47:42 2015-01-09 18:34:37   
3843 2015-01-07 19:53:07 2015-01-08 19:40:45 2015-01-09 19:14:16   
4277 2015-01-07 18:07:27 2015-01-08 18:14:43 2015-01-09 17:44:55   

              2015-01-12          2015-01-13          2015-01-15  ...  \
1353 2015-01-12 17:54:28 2015-01-13 17:31:12 2015-01-15 17:07:51  ...   
994  2015-01-12 17:26:16 2015-01-13 19:13:08 2015-01-15 19:41:33  ...   
1476 2015-01-12 16:13:19 2015-01-13 18:00:20 2015-01-15 17:27:41  ...   
3437 2015-01-12 17:14:57 2015-01-13 17:09:49 2015-01-15 17:45:43  ...   
4397 2015-01-12 16:48:38 2015-01-13 17:02:15 2015-01-15 17:02:12  ...   
813  2015-01-12 17:19:42 2015-01-13 16:15:48 2015-01-15 16:11:04  ...   
4204 2015-01-12 18:55:43 2015-01-13 19:16:58 2015-01-15 19:31:02  ...   
2114 2015-01-12 18:02:31 2015-01-13 17:44:29 2015-01-15 17:36:27  ...   
3843 2015-01-12 19:51:42 2015-01-13 19:27:13 2015-01-15 19:12:41  ...   
4277 2015-01-12 18:14:21 2015-01-13 17:21:17 2015-01-15 17:23:40  ...   

              2015-12-17          2015-12-18          2015-12-21  \
1353 2015-12-17 17:27:51 2015-12-18 20:08:21 2015-12-21 17:38:22   
994  2015-12-17 19:44:58 2015-12-18 20:41:59 2015-12-21 19:51:55   
1476 2015-12-17 17:03:19 2015-12-18 16:53:36 2015-12-21 16:54:05   
3437 2015-12-17 17:18:29 2015-12-18 16:46:27 2015-12-21 16:55:12   
4397 2015-12-17 18:08:05 2015-12-18 16:39:33 2015-12-21 17:22:39   
813  2015-12-17 16:26:29 2015-12-18 16:16:39 2015-12-21 17:09:21   
4204 2015-12-17 20:08:06 2015-12-18 19:36:59 2015-12-21 19:03:55   
2114 2015-12-17 19:01:56 2015-12-18 18:24:04 2015-12-21 17:42:28   
3843 2015-12-17 19:58:12 2015-12-18 19:34:10 2015-12-21 19:15:41   
4277 2015-12-17 17:06:41 2015-12-18 16:37:15 2015-12-21 18:04:12   

              2015-12-22          2015-12-23          2015-12-24  \
1353 2015-12-22 17:40:04 2015-12-23 16:39:17 2015-12-24 18:22:49   
994  2015-12-22 20:09:47 2015-12-23 19:28:56 2015-12-24 19:46:03   
1476 2015-12-22 16:36:07 2015-12-23 16:48:26 2015-12-24 16:50:49   
3437 2015-12-22 18:00:47 2015-12-23 17:36:43 2015-12-24 17:15:10   
4397 2015-12-22 17:27:28 2015-12-23 17:28:02 2015-12-24 16:55:28   
813  2015-12-22 17:04:05 2015-12-23 16:44:34 2015-12-24 16:49:45   
4204 2015-12-22 19:07:02 2015-12-23 19:06:11 2015-12-24 19:39:19   
2114 2015-12-22 18:14:38 2015-12-23 17:29:05 2015-12-24 17:14:53   
3843 2015-12-22 19:29:34 2015-12-23 19:27:19 2015-12-24 19:14:15   
4277 2015-12-22 17:17:02 2015-12-23 17:07:00 2015-12-24 18:08:24   

              2015-12-28          2015-12-29          2015-12-30  \
1353 2

<p>
    Pour réduire la dimension des données, nous allons créer une nouvelle variable représentant la moyenne des horaires de chaque employé. 
    Cela sera réalisé en calculant la moyenne des horaires pour chaque ligne à partir des colonnes existantes dans les DataFrames 'in_time' et 'out_time'.
  </p>

In [69]:
# Calculer l'horaire moyen d'entrée pour chaque employé dans 'in_time'
in_time = average_time(in_time, in_time.columns[1:], 'average_start_time')

e:\Sami\Sami\Sami\ISIMa\2 TSD\Data_Analysis_Python\HumanForYou\python_project\python_project\work_tools.py:332: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  dataframe[column_name] = average_start_time


In [70]:
in_time['average_start_time']

0       09:59:50.120481
1       09:59:05.542168
2       10:00:47.228915
3       09:58:10.602409
4       09:59:14.939759
             ...       
4405    10:00:43.132530
4406    09:59:46.265060
4407    09:59:06.506024
4408    10:00:36.626506
4409    10:00:45.783132
Name: average_start_time, Length: 4410, dtype: object

In [71]:
# Calculer l'horaire moyen d'entrée pour chaque employé dans 'out_time'
out_time = average_time(out_time,out_time.columns[1:], 'average_end_time')

e:\Sami\Sami\Sami\ISIMa\2 TSD\Data_Analysis_Python\HumanForYou\python_project\python_project\work_tools.py:332: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  dataframe[column_name] = average_start_time


In [72]:
out_time['average_end_time']

0       17:23:34.939759
1       17:39:25.542168
2       17:02:25.783132
3       17:12:52.289156
4       18:02:24.096385
             ...       
4405    18:29:15.903614
4406    16:08:14.939759
4407    17:49:15.662650
4408    19:25:25.542168
4409    17:10:06.265060
Name: average_end_time, Length: 4410, dtype: object

Création d'un DataFrame Consolidé 'in_out_time'

In [73]:
# Créer un DataFrame consolidé 'in_out_time' avec les colonnes pertinentes
in_out_time = pd.concat([in_time['Unnamed: 0'], in_time['average_start_time'], out_time['average_end_time']], axis=1)

In [74]:
# Renommer la colonne 'Unnamed: 0' en 'EmployeeID' dans le DataFrame 'in_out_time'
in_out_time = in_out_time.rename(columns={'Unnamed: 0': 'EmployeeID'})

In [75]:
# Coup d'oeil
in_out_time

EmployeeID average_start_time average_end_time
0              1    09:59:50.120481  17:23:34.939759
1              2    09:59:05.542168  17:39:25.542168
2              3    10:00:47.228915  17:02:25.783132
3              4    09:58:10.602409  17:12:52.289156
4              5    09:59:14.939759  18:02:24.096385
...          ...                ...              ...
4405        4406    10:00:43.132530  18:29:15.903614
4406        4407    09:59:46.265060  16:08:14.939759
4407        4408    09:59:06.506024  17:49:15.662650
4408        4409    10:00:36.626506  19:25:25.542168
4409        4410    10:00:45.783132  17:10:06.265060

[4410 rows x 3 columns]

<p>
    Pour faciliter les calculs et analyses liés au temps, nous effectuons les étapes suivantes :
  </p>

  <ol>
    <li>
      <p>Convertir les colonnes 'average_start_time' et 'average_end_time' en timedelta :</p>
      <pre>
        in_out_time['average_start_time'] = pd.to_timedelta(in_out_time['average_start_time'].apply(lambda x: str(x)))
        in_out_time['average_end_time'] = pd.to_timedelta(in_out_time['average_end_time'].apply(lambda x: str(x)))
      </pre>
    </li>
    <li>
      <p>Calculer la différence entre les deux colonnes :</p>
      <pre>
        in_out_time['difference_temps'] = in_out_time['average_end_time'] - in_out_time['average_start_time']
      </pre>
    </li>
  </ol>

  <p>
    Ces étapes visent à transformer les données temporelles dans le DataFrame 'in_out_time', permettant ainsi le calcul de la durée totale du temps de travail pour chaque employé.
  </p>

In [76]:
# Convertir les colonnes 'average_start_time' et 'average_end_time' en timedelta
in_out_time['average_start_time'] = pd.to_timedelta(in_out_time['average_start_time'].apply(lambda x: str(x)))
in_out_time['average_end_time'] = pd.to_timedelta(in_out_time['average_end_time'].apply(lambda x: str(x)))

# Calculer la différence entre les deux colonnes
in_out_time['difference_temps'] = in_out_time['average_end_time'] - in_out_time['average_start_time']

In [77]:
in_out_time

EmployeeID     average_start_time       average_end_time  \
0              1 0 days 09:59:50.120481 0 days 17:23:34.939759   
1              2 0 days 09:59:05.542168 0 days 17:39:25.542168   
2              3 0 days 10:00:47.228915 0 days 17:02:25.783132   
3              4 0 days 09:58:10.602409 0 days 17:12:52.289156   
4              5 0 days 09:59:14.939759 0 days 18:02:24.096385   
...          ...                    ...                    ...   
4405        4406 0 days 10:00:43.132530 0 days 18:29:15.903614   
4406        4407 0 days 09:59:46.265060 0 days 16:08:14.939759   
4407        4408 0 days 09:59:06.506024 0 days 17:49:15.662650   
4408        4409 0 days 10:00:36.626506 0 days 19:25:25.542168   
4409        4410 0 days 10:00:45.783132 0 days 17:10:06.265060   

           difference_temps  
0    0 days 07:23:44.819278  
1           0 days 07:40:20  
2    0 days 07:01:38.554217  
3    0 days 07:14:41.686747  
4    0 days 08:03:09.156626  
...                     ...  
4405 0 days 08:28:32.771084  
4406 0 days 06:08:28.674699  
4407 0 days 07:50:09.156626  
4408 0 days 09:24:48.915662  
4409 0 days 07:09:20.481928  

[4410 rows x 4 columns]

In [78]:
# Fonction pour formater les colonnes de temps en HH:MM:SS.mmmmmm
def format_timedelta_column(column):
    column_seconds = column.total_seconds()
    hours, remainder = divmod(column_seconds, 3600)
    minutes, seconds = divmod(remainder, 60)
    return "{:02}:{:02}:{:06.6f}".format(int(hours), int(minutes), seconds)

# Appliquer la fonction pour formater toutes les colonnes de temps
time_columns = ['average_start_time', 'average_end_time', 'difference_temps']
for col in time_columns:
    in_out_time[col] = in_out_time[col].apply(format_timedelta_column)

Cette fonction et le code qui l'accompagne sont utilisés pour formater les colonnes de temps dans le DataFrame 'in_out_time' afin d'obtenir une représentation plus lisible et conviviale.

In [79]:
in_out_time

EmployeeID average_start_time average_end_time difference_temps
0              1    09:59:50.120481  17:23:34.939759  07:23:44.819278
1              2     09:59:5.542168  17:39:25.542168  07:40:20.000000
2              3    10:00:47.228915  17:02:25.783132  07:01:38.554217
3              4    09:58:10.602409  17:12:52.289156  07:14:41.686747
4              5    09:59:14.939759  18:02:24.096385   08:03:9.156626
...          ...                ...              ...              ...
4405        4406    10:00:43.132530  18:29:15.903614  08:28:32.771084
4406        4407    09:59:46.265060  16:08:14.939759  06:08:28.674699
4407        4408     09:59:6.506024  17:49:15.662650   07:50:9.156626
4408        4409    10:00:36.626506  19:25:25.542168  09:24:48.915662
4409        4410    10:00:45.783132   17:10:6.265060  07:09:20.481928

[4410 rows x 4 columns]

### **Export Dataset**
Exportation du DataFrame 'in_out_time' vers un fichier CSV

In [80]:
# Exporter le DataFrame 'in_out_time' vers un fichier CSV
in_out_time.to_csv('./in_out_time.csv', index=False)